decompress_tar.go

```diff
@@ -45,7 +58,15 @@ func untar(input io.Reader, dst, src string, dir bool, umask os.FileMode) error
 			path = filepath.Join(path, hdr.Name)
 		}
 
-		if hdr.FileInfo().IsDir() {
+		fileInfo := hdr.FileInfo()
+
+		fileSize += fileInfo.Size()
+
+		if fileSizeLimit > 0 && fileSize > fileSizeLimit {
+			return fmt.Errorf("tar archive larger than limit: %d", fileSizeLimit)
+		}
```

Comment:
> The approach here (and in decompress_zip.go) seems like we trust that the headers for the archive can't lie to us about the size of the file. Is that actually guaranteed? I can't figure out whether that would corrupt the header to the point where it'd just blow up trying to decompress anyways, but I've admittedly not spent a ton of time on that.

Not trusting the headers would be preferable, but it's also unclear to me if it's possible to tamper them like that. I can spend some time trying to figure that out.

> Threading the limit and cumulative file sizes into the copyReader seems like it could avoid that risk and let us share the implementation between all consumers of that function.

Fair point. There's a little more to unwind there, I think: `copyReader` is specifically used to copy a _single_ file (`io.Reader`). I added a `fileSizeLimit` function parameter, so when given a value greater than zero, it wraps the given file reader in an `io.LimitReader`, which has its own limitations (returning `EOF`).

https://github.com/hashicorp/go-getter/blob/cf15d8405d9a700e9c25df7194d095b7dfdff914/get_file_copy.go#L34

https://github.com/hashicorp/go-getter/blob/cf15d8405d9a700e9c25df7194d095b7dfdff914/get_file_copy.go#L41-L43

So, would you see an _additional parameter_ being used here for cumulative files? Or, how would you see threading these limits together in general? 🤔

Reply:
> So, would you see an _additional parameter_ being used here for cumulative files? Or, how would you see threading these limits together in general?

Something like that. Or have the `copyReader` take a pointer to an integer that represents the remaining limit, that it decrements by the bytes read. But writing it out like that makes it seem like it might be too complex of an implementation for this fix. Especially if it turns out the headers can be reasonably trusted anyways.

Reply:
I guess it boils down to "are these _user supplied_ archive headers trustable" before messing with the `copyReader` any further.

Reply:
Looking at ZIP first, I found an interesting related issue I figured I'd link here for context: https://github.com/golang/go/issues/33036

Reply:
I went to the Go source and it looks like they specifically check that they're not reading more than the compressed size: https://cs.opensource.google/go/go/+/master:src/archive/zip/reader.go;l=293

Reply:
The `archive/tar` package does as well: https://cs.opensource.google/go/go/+/master:src/archive/tar/reader.go;l=44-45;bpv=0

So I think we're actually safe here. In which case I'm :+1: to resolve this thread.